test/SNLS_testdriver_hybrd.cc

```diff
@@ -0,0 +1,368 @@
+  ACM TOMS, Vol. 7, No. 1, March 1981, pp. 14-41.  The modification
+  involves squaring the last equation fn(x) and using it in a
+  homotopy-type equation.
+
+  The parameter "lambda" is a homotopy-type parameter that may be
+  varied from 0 to 1 to adjust the ill-conditioning of the problem.
+  A value of 0 is the original, unmodified problem, while a value of
+  1 is that problem with the last equation squared.  Typical values
+  for increasingly ill-conditioned problems might be 0.9, 0.99,
+  0.999, etc.
+
+  The standard starting point is x(i) = -1, but setting x(i) = 0 tests
+  the selected global strategy.
+*/
+class Broyden
+{
+public:
+   static const int nDimSys = 8 ;
+
+   // constructor
+   __snls_hdev__  Broyden(double lambda )
+      : _lambda(lambda)
+      {
+         std::cout << "Broyden ill-conditioning: lambda = "
+                   << std::setw(21) << std::setprecision(11) << _lambda << std::endl;          
+      } ;
+
+   __snls_hdev__ bool computeRJ(double* const r,
+                                double* const J,
+                                const double* const x )
+      {
+         double fn ;
+         const int nDim = nDimSys ; // convenience -- less code change below
+         
+#if DEBUG > 1
+         std::cout << "Evaluating at x = " ;
+         for (int i=1; i<nDim; ++i) {
+            std::cout << std::setw(21) << std::setprecision(11) << x[i] << " ";
+         }
+         std::cout << std::endl ;
+#endif
+
+         bool doComputeJ = (J != nullptr) ;
+         if ( doComputeJ ) {
+            for ( int ijJ=0; ijJ<_nXn; ++ijJ ) {
+               J[ijJ] = 0.0 ;
+            }
+         }
+         
+         r[0] = (3-2*x[0])*x[0] - 2*x[1] + 1;
+         for (int i=1; i<nDim-1; i++)
+            r[i] = (3-2*x[i])*x[i] - x[i-1] - 2*x[i+1] + 1;
+
+         fn = (3-2*x[nDim-1])*x[nDim-1] - x[nDim-2] + 1;
+         r[nDim-1] = (1-_lambda)*fn + _lambda*(fn*fn);
+
+         if ( doComputeJ ) {
+            // F(0) = (3-2*x[0])*x[0] - 2*x[1] + 1;
+            J[SNLS_NN_INDX(0,0,nDim)] = 3 - 4*x[0];
+            J[SNLS_NN_INDX(0,1,nDim)] = -2;
+
+            // F(i) = (3-2*x[i])*x[i] - x[i-1] - 2*x[i+1] + 1;
+            for (int i=1; i<nDim-1; i++) {
+               J[SNLS_NN_INDX(i,i-1,nDim)] = -1;
+               J[SNLS_NN_INDX(i,i,nDim)]   = 3 - 4*x[i];
+               J[SNLS_NN_INDX(i,i+1,nDim)] = -2;
+            }
+
+            // F(n-1) = ((3-2*x[n-1])*x[n-1] - x[n-2] + 1)^2;
+            fn = (3-2*x[nDim-1])*x[nDim-1] - x[nDim-2] + 1;
+            double dfndxn = 3-4*x[nDim-1];
+            J[SNLS_NN_INDX(nDim-1,nDim-1,nDim)] = (1-_lambda)*(dfndxn) + _lambda*(2*dfndxn*fn);
+            J[SNLS_NN_INDX(nDim-1,nDim-2,nDim)] = (1-_lambda)*(-1) + _lambda*(-2*fn);
+         }
+
+         return true ;
+         
+      } ;
+   
+   private:
+      double _lambda ;
+      static const int _nXn = nDimSys*nDimSys ;
+};
+
+template<int nd>
+class ChebyQuad
```

Comment:
Is there a reference or other documentation to be added for this test problem?

Reply:
I do have a few references for this. I actually found the original paper this was defined in, but I found this in a nonlinear solver test suite elsewhere that I can also reference.